hollow/src/main/java/com/netflix/hollow/api/consumer/index/PrimaryKeyIndex.java

```diff
@@ -0,0 +1,319 @@
+package com.netflix.hollow.api.consumer.index;
+
+import static java.util.stream.Collectors.toList;
+
+import com.netflix.hollow.api.consumer.HollowConsumer;
+import com.netflix.hollow.api.custom.HollowAPI;
+import com.netflix.hollow.api.objects.HollowObject;
+import com.netflix.hollow.core.index.FieldPaths;
+import com.netflix.hollow.core.index.HollowPrimaryKeyIndex;
+import com.netflix.hollow.core.index.key.PrimaryKey;
+import com.netflix.hollow.core.read.engine.HollowReadStateEngine;
+import com.netflix.hollow.core.schema.HollowObjectSchema;
+import com.netflix.hollow.core.schema.HollowSchema;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * A type safe index for indexing with a primary, or unique, key.
+ *
+ * @param <T> the primary type
+ * @param <Q> the key type
+ */
+public class PrimaryKeyIndex<T extends HollowObject, Q> {
+    final HollowConsumer consumer;
+    HollowAPI api;
+    boolean listenToDataRefresh;
+    final SelectFieldPathResultExtractor<T> primaryTypeExtractor;
+    final List<MatchFieldPathArgumentExtractor<Q>> matchFields;
+    final String primaryTypeName;
+    final String[] matchFieldPaths;
+    HollowPrimaryKeyIndex hpki;
+    RefreshListener refreshListener;
+
+    PrimaryKeyIndex(
+            HollowConsumer consumer,
+            Class<T> primaryType,
+            PrimaryKey primaryTypeKey,
+            boolean listenToDataRefresh,
```

Comment:
Maybe it's okay to remove the constructor argument and require that users must call the method.